contributors/design-proposals/hpa-v2.md

```diff
@@ -0,0 +1,291 @@
+)
+
+// a specification for how to scale based on a single metric
+// (only `type` and one other matching field should be set at once)
+type MetricSpec struct {
+    // the type of metric source (should match one of the fields below)
+    Type MetricSourceType `json:"type"`
+
+    // a metric describing a single kubernetes object (for example, hits-per-second on an Ingress object)
+    Object *ObjectMetricSource `json:"object,omitempty"`
+    // a metric describing each pod in the current scale target (for example, transactions-processed-per-second).
+    // The values will be averaged together before being compared to the target value
+    Pods *PodsMetricSource `json:"pods,omitemtpy"`
+    // a resource metric (such as those specified in requests and limits) known to Kubernetes
+    // describing each pod in the current scale target (e.g. CPU or memory). Such metrics are
+    // built in to Kubernetes, and have special scaling options on top of those available to
+    // normal per-pod metrics using the "pods" source.
+    Resource *ResourceMetricSource `json:"resource,omitempty"`
+}
+
+// a metric describing a single kubernetes object (for example, hits-per-second on an Ingress object)
+type ObjectMetricSource struct {
+    // the described Kubernetes object
+    Target CrossVersionObjectReference `json:"target"`
+
+    // the name of the metric in question
+    MetricName string `json:"metricName"`
+    // the target value of the metric (as a quantity)
+    TargetValue resource.Quantity `json:"targetValue"`
+}
+
+// a metric describing each pod in the current scale target (for example, transactions-processed-per-second).
+// The values will be averaged together before being compared to the target value
+type PodsMetricSource struct {
+    // the name of the metric in question
+    MetricName string `json:"metricName"`
+    // the target value of the metric (as a quantity)
+    TargetAverageValue resource.Quantity `json:"targetAverageValue"`
+}
+
+// a resource metric known to Kubernetes, as specified in requests and limits, describing each pod
+// in the current scale target (e.g. CPU or memory).  The values will be averaged together before
+// being compared to the target.  Such metrics are built in to Kubernetes, and have special
+// scaling options on top of those available to normal per-pod metrics using the "pods" source.
+// Only one "target" type should be set.
+type ResourceMetricSource struct {
+    // the name of the resource in question
+    Name api.ResourceName `json:"name"`
+    // the target value of the resource metric, represented as
+    // a percentage of the requested value of the resource on the pods.
+    // +optional
+    TargetAverageUtilization *int32 `json:"targetAverageUtilization,omitempty"`
+    // the target value of the resource metric as a raw value, similarly
+    // to the "pods" metric source type.
+    // +optional
+    TargetAverageRawValue *resource.Quantity `json:"targetAverageRawValue,omitempty"`
```

Comment:
I'd suggest keeping it consistent with `PodMetricSource`, i.e. `TargetAverageValue`.